// @modzy/modzy-sdk/dist/utils.js
"use strict";

function _awaitAsyncGenerator(value) { return new _AwaitValue(value); }

function _wrapAsyncGenerator(fn) { return function () { return new _AsyncGenerator(fn.apply(this, arguments)); }; }

function _AsyncGenerator(gen) { var front, back; function send(key, arg) { return new Promise(function (resolve, reject) { var request = { key: key, arg: arg, resolve: resolve, reject: reject, next: null }; if (back) { back = back.next = request; } else { front = back = request; resume(key, arg); } }); } function resume(key, arg) { try { var result = gen[key](arg); var value = result.value; var wrappedAwait = value instanceof _AwaitValue; Promise.resolve(wrappedAwait ? value.wrapped : value).then(function (arg) { if (wrappedAwait) { resume(key === "return" ? "return" : "next", arg); return; } settle(result.done ? "return" : "normal", arg); }, function (err) { resume("throw", err); }); } catch (err) { settle("throw", err); } } function settle(type, value) { switch (type) { case "return": front.resolve({ value: value, done: true }); break; case "throw": front.reject(value); break; default: front.resolve({ value: value, done: false }); break; } front = front.next; if (front) { resume(front.key, front.arg); } else { back = null; } } this._invoke = send; if (typeof gen.return !== "function") { this.return = undefined; } }

if (typeof Symbol === "function" && Symbol.asyncIterator) { _AsyncGenerator.prototype[Symbol.asyncIterator] = function () { return this; }; }

_AsyncGenerator.prototype.next = function (arg) { return this._invoke("next", arg); };

_AsyncGenerator.prototype.throw = function (arg) { return this._invoke("throw", arg); };

_AsyncGenerator.prototype.return = function (arg) { return this._invoke("return", arg); };

function _AwaitValue(value) { this.wrapped = value; }

const fs = require('fs');

function readBytes(fd, sharedBuffer) {
  return new Promise((resolve, reject) => {
    fs.read(fd, sharedBuffer, 0, sharedBuffer.length, null, err => {
      if (err) {
        return reject(err);
      }

      resolve();
    });
  });
}

function fileToChunks(_x, _x2) {
  return _fileToChunks.apply(this, arguments);
}

function _fileToChunks() {
  _fileToChunks = _wrapAsyncGenerator(function* (filePath, size) {
    const sharedBuffer = Buffer.alloc(size);
    const stats = fs.statSync(filePath);
    const fd = fs.openSync(filePath, "r");
    let bytesRead = 0;
    let end = size;

    for (let i = 0; i < Math.ceil(stats.size / size); i++) {
      yield _awaitAsyncGenerator(readBytes(fd, sharedBuffer));
      bytesRead = (i + 1) * size;

      if (bytesRead > stats.size) {
        end = size - (bytesRead - stats.size);
      }

      yield sharedBuffer.slice(0, end);
    }

    fs.closeSync(fd);
  });
  return _fileToChunks.apply(this, arguments);
}

function byteArrayToChunks(_x3, _x4) {
  return _byteArrayToChunks.apply(this, arguments);
}

function _byteArrayToChunks() {
  _byteArrayToChunks = _wrapAsyncGenerator(function* (byteArray, size) {
    for (let i = 0; i < byteArray.length; i += size) {
      yield byteArray.slice(i, i + size);
    }
  });
  return _byteArrayToChunks.apply(this, arguments);
}

module.exports = {
  byteArrayToChunks,
  fileToChunks
};
